Add "getting started" docs (#306)
namely/chief-of-state: docs/getting-started.md

```

From there, you can generate the COS interfaces for your language (see the
official quick start guide for your language https://grpc.io/docs/languages/).

As an example, one would use [protoc](https://grpc.io/docs/protoc-installation/)
to generate java source code like so:
```sh
export SRC_DIR=protos/cos
export DST_DIR=build/gen
# generate everything in protos/cos/chief_of_state/v1/common.proto
protoc -I=$SRC_DIR --java_out=$DST_DIR $SRC_DIR/foo.proto
# generate ALL .proto files in protos/cos
find $SRC_DIR -name "*.proto" | xargs -I{} protoc -I=$SRC_DIR --java_out=$DST_DIR {}
```

**Pro tip:** many languages have wrappers around protoc to make this easier,
like [scalap-pb](https://scalapb.github.io/) for scala (which chief-of-state
uses internally). Namely also offers an open-source docker-based solution
[namely/docker-protoc](https://github.com/namely/docker-protoc) with support
for many languages.

## Implement a "Write Side"

Drawing from CQRS terminology, COS defines a [writeside interface](https://github.com/namely/chief-of-state-protos/blob/master/chief_of_state/v1/writeside.proto) that you must implement to define how state
is created and updated.

### Methods

> `HandleCommand` accepts a command (think "request") and a prior state and returns an event.

Reviewer: nit: 
```suggestion
`HandleCommand` accepts a command (or a "request" if you will) and a prior state and returns an event.
```

Author: don't say i never did nothin for ya

For example, given a command to `UpdateUserEmail` and a User, this RPC might return

> an event called `UserEmailUpdated`. This method is allowed to throw exceptions
> (as gRPC error statuses) if the inbound command (request) is invalid.

Reviewer: ```suggestion
an event called `UserEmailUpdated`. This method is encouraged to throw exceptions
(as gRPC error statuses) **if** the inbound command (request) is invalid.
```

Author: don't say i never did nothin for ya

`HandleEvent` accepts an event and the prior state of an entity and returns a new state.

> For example, given a UserEmailUpdated event and a User, this RPC would return a

Reviewer: ```suggestion
For example, given a `UserEmailUpdated` event and a User, this RPC would return a
```

new User instance with the email updated. It's advisable to do all your
validations in `HandleCommand` such that this method is unlikely to encounter
invalid data. If `HandleCommand` fails, `HandleEvent` is not called for that
command.

> ### Registration

Reviewer: maybe we have this explained somewhere else but I would consider reiterate what is explained in the diagram and then how cos requires some mandatory vars with the write handler vars among them.

Author: A few lines down there's a link to the full config doc.

Regarding the diagram, yeah maybe that and a few sentences describing the components at the top. I'm torn about the scope of this doc.

Register your write handler service as COS environment variables like so:

```yaml
# set the host and port for your write handler gRPC server
COS_WRITE_SIDE_HOST: localhost
COS_WRITE_SIDE_PORT: 50051
```

(If you are new to setting up COS, see the full [configuration](docs/configuration.md) guide.)

### Sending requests to COS

> Once you have implemneted a write handler and registered it with COS, you are

Reviewer: ```suggestion
Once you have implemented a write handler and registered it with COS, you are
```

able to start sending requests to chief-of-state with the chief-of-state
[gRPC client methods](https://github.com/namely/chief-of-state-protos/blob/master/chief_of_state/v1/service.proto).

> `ProcessCommand` accepts a command (request) and an entity ID, and processes

Reviewer: I think we should reiterate what the entity ID is.
I like the explanation for GetState to think of this as keys and values and I think it's also useful for the context of creating and updating an entity.

that command by forwarding it to your write handler methods. Commands for a given
entity are guaranteed to be processed serially, but commands for unrelated entities
can be processed in parallel.

`GetState` accepts an entity ID and returns the latest state of that entity. Use
this like a key/value lookup.

## Implement a "Read Side"

COS defines a [readside interface](https://github.com/namely/chief-of-state-protos/blob/master/chief_of_state/v1/readside.proto)
and allows you to implement many "readsides" (from CQRS "read side") to
stream your events and state to external destinations. For example, you might
implement a readside that writes to your applications private elasticsearch
cluster for serving fancy search queries, and you could implement a second
readside that publishes your event to kafka for consumption by other services.

The COS readside processor guarantees that events are served in the order they were

> persisted to the journal per entity. Offsets are tracked for per readside, so

Reviewer: I think we should link or add a few words explaining what a journal is, as part of  the cqrs world

Author: Yeah, I wrestled with how in depth this doc should be.

Reviewer 2: I don't think we should explain what Journal is in the CQRS world. There are a lot of doc out there explaining. We can just elaborate that COS help build event sourced/cqrs driven application and maybe and I mean maybe point to the some docs.
